### DISCUSSION TITLE
Need help to print url from route

### DISCUSSION THREAD

**Discussion by trispo** [pointfreeco/swift-url-routing#65] [Q&A] [answered]
Hi,

I need some help on printing a url from my route. 

```swift
enum AppRoute {
    case foo(String)
}

let router = OneOf {
    Route(AppRoute.foo) {
        Path {
            "foo"
            Parse(.string)
        }
    }
}

guard let url = URLRequest(
    data: try router
        .baseURL("https://my-host.app")
        .print(.foo("barbaz"))
)?.url else { return }
```

With the code above, I get the following compilation error:
`Referencing instance method 'baseURL' on 'Route' requires that 'Parsers.Map<Path<PathBuilder.PathZipVO<String, Parse<Parsers.MapConversion<Rest<Substring>, Conversions.SubstringToString>>>>, AppRoute>' conform to 'ParserPrinter'`

How can I get this to compile?

Thanks
@trispo

**Comment by saroar** [comment]
I am not sure it will help you or not try check lib tests

**Comment by trispo** [comment]
Unfortunately no. The samples there are too simple. They don't use a route with associated value. Without the associated string the compiler is happy. But I need that value.

**Reply by saroar** [reply]
can you try like this 
enum AppRoute {
    case foo(String)
}

let router = OneOf {
    Route(AppRoute.foo) {
        Path {
            "foo"
            Parse(.string)
        }
    }
}

let apiRouter = router.baseURL("http://10.0.1.4:7070")

let hihi = router.url(for: .foo(""))
debugPrint(hihi)
also note `invitation` is not `foo`
<img width="336" alt="Screenshot 2022-11-24 at 18 09 08" src="https://user-images.githubusercontent.com/8770772/203816864-db51344e-3c12-4151-867e-61c509f351dd.png">

**Reply by trispo** [reply]
Well, this doesn't compile either. The "invitation" route was a leftover. I renamed (nearly) everything for this thread. I've updated the original post.

**Reply by mmllr** [reply]
You forgot the .case in the router:

```swift
enum TestRoute {
    case foo(String)
}

let theRouter = OneOf {
    Route(.case(TestRoute.foo)) {
            Path {
                "foo"
                Parse(.string)
            }
        }
}

let url = try theRouter
        .baseURL("https://my-host.app")
        .url(for: .foo("1234"))
```

**Reply by trispo** [reply]
Great. This works. Thank you so much.

**Reply by trispo** [reply]
@mmllr Would you please copy your reply to the "Suggest an answer" box so that I can mark it as the accepted answer?

**Comment by mmllr** [comment, marked as answer]
You need to add the ```.case``` in the router to make it a ParserPrinter:

```swift
enum TestRoute {
    case foo(String)
}

let theRouter = OneOf {
    Route(.case(TestRoute.foo)) {
            Path {
                "foo"
                Parse(.string)
            }
        }
}

let url = try theRouter
        .baseURL("https://my-host.app")
        .url(for: .foo("1234"))
```
